File: 462-minimum-moves-to-equal-array-elements-ii/462-minimum-moves-to-equal-array-elements-ii.c


int comp(const void* a , const void* b) {
    return *((int*)a) - *((int*)b);
}

int minMoves2(int* n, int S){
    
    if (S == 1) return 0; 
    
    qsort(n, S, sizeof(int), comp);
    int i;
    /*for(i = 0; i < S; i++) {
        printf("%d ", n[i]);
    }
    printf("\n");*/
    
    int mid = n[S/2];
    int min = 0;
    for(i = 0; i < S/2; i++) {
        min += mid - n[i];
    }
    for(i = S/2 + 1; i < S; i++) {
        min += n[i] - mid;
    }
    //min -= mid*(S - S/2 -1);
    
    return min;
        
}

File: 462-minimum-moves-to-equal-array-elements-ii/462-minimum-moves-to-equal-array-elements-ii.py
class Solution(object):
    def minMoves2(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """  
        
        nums.sort()
        
        ltar = nums[0]
        rtar = nums[-1]
        
        c = [0]*len(nums)
        
        lprev = 0
        rprev = 0
        for i in xrange(len(nums)):
            ri = len(nums) - 1 - i
            
            la = (nums[i] - ltar)*i 
            ra = (-nums[ri] + rtar)*i 
            c[i] += la + lprev
            c[ri] += ra + rprev
            
            lprev += la
            rprev += ra
                     
            ltar = nums[i]
            rtar = nums[ri]
            
            #print(nums[i], nums[ri], c)
            
        return min(c)
            
            
            
        
        
        
        
        
            
            
            
            